File: labs/M18-evaluation-testing/solution/task_scorer.py
"""
M18 — Task Completion Scorer (Solution)
=========================================
Scores whether the agent found the correct UCC filings.
Uses partial credit: finding 3 of 5 expected filings = 0.6.
"""

import re


def extract_filing_numbers(text: str) -> list[str]:
    """
    Extract UCC filing numbers from response text.
    Pattern: UCC-YYYY-XX-NNNNNNN (e.g., UCC-2024-NY-0012847)
    """
    pattern = r"UCC-\d{4}-[A-Z]{2}-\d{7}"
    matches = re.findall(pattern, text)
    return list(set(matches))  # deduplicate


def score_task_completion(response: str, expected: dict) -> dict:
    """
    Score whether the agent found the correct filings.

    Returns:
        {"score": 0.0-1.0, "found": [...], "missed": [...], "extra": [...], "details": str}
    """
    expected_filings = set(expected.get("expected_filings", []))
    found_filings = set(extract_filing_numbers(response))

    # Edge case: no filings expected
    if not expected_filings:
        if not found_filings:
            return {
                "score": 1.0,
                "found": [],
                "missed": [],
                "extra": [],
                "details": "Correctly returned no filings (none expected).",
            }
        else:
            return {
                "score": 0.0,
                "found": [],
                "missed": [],
                "extra": sorted(found_filings),
                "details": f"Expected no filings but found {len(found_filings)}.",
            }

    # Calculate matches
    correctly_found = expected_filings & found_filings
    missed = expected_filings - found_filings
    extra = found_filings - expected_filings

    # Partial credit: proportion of expected filings that were found
    score = len(correctly_found) / len(expected_filings)

    details_parts = []
    if correctly_found:
        details_parts.append(f"Found {len(correctly_found)}/{len(expected_filings)} expected filings.")
    if missed:
        details_parts.append(f"Missed: {', '.join(sorted(missed))}.")
    if extra:
        details_parts.append(f"Extra (not expected): {', '.join(sorted(extra))}.")
    if score == 1.0:
        details_parts.append("Perfect match!")

    return {
        "score": score,
        "found": sorted(correctly_found),
        "missed": sorted(missed),
        "extra": sorted(extra),
        "details": " ".join(details_parts),
    }


# ---------------------------------------------------------------------------
# Self-test
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    print("Task Completion Scorer — Self-Test")
    print("=" * 50)

    # Test 1: Perfect match
    response1 = (
        "Found 2 filings:\n"
        "- UCC-2024-NY-0012847 (Active)\n"
        "- UCC-2024-NY-0015921 (Active)"
    )
    expected1 = {
        "expected_filings": ["UCC-2024-NY-0012847", "UCC-2024-NY-0015921"]
    }
    result1 = score_task_completion(response1, expected1)
    print(f"\nTest 1 — Perfect match:")
    print(f"  Score: {result1['score']} (expected: 1.0)")
    print(f"  Found: {result1['found']}")
    print(f"  Missed: {result1['missed']}")
    assert result1["score"] == 1.0, f"Expected 1.0, got {result1['score']}"

    # Test 2: Partial match (1 of 2)
    response2 = "I found filing UCC-2024-NY-0012847 for Acme Corporation."
    expected2 = {
        "expected_filings": ["UCC-2024-NY-0012847", "UCC-2024-NY-0015921"]
    }
    result2 = score_task_completion(response2, expected2)
    print(f"\nTest 2 — Partial match (1/2):")
    print(f"  Score: {result2['score']} (expected: 0.5)")
    print(f"  Found: {result2['found']}")
    print(f"  Missed: {result2['missed']}")
    assert result2["score"] == 0.5, f"Expected 0.5, got {result2['score']}"

    # Test 3: No match
    response3 = "I could not find any filings for that entity."
    expected3 = {
        "expected_filings": ["UCC-2024-NY-0012847"]
    }
    result3 = score_task_completion(response3, expected3)
    print(f"\nTest 3 — No match:")
    print(f"  Score: {result3['score']} (expected: 0.0)")
    assert result3["score"] == 0.0, f"Expected 0.0, got {result3['score']}"

    # Test 4: Empty expected
    response4 = "No filings were found for XYZ Corp."
    expected4 = {"expected_filings": []}
    result4 = score_task_completion(response4, expected4)
    print(f"\nTest 4 — Empty expected, no filings in response:")
    print(f"  Score: {result4['score']} (expected: 1.0)")
    assert result4["score"] == 1.0, f"Expected 1.0, got {result4['score']}"

    # Test 5: Extra filings found
    response5 = (
        "Found 3 filings:\n"
        "- UCC-2024-NY-0012847\n"
        "- UCC-2024-NY-0015921\n"
        "- UCC-2024-CA-0101457"
    )
    expected5 = {
        "expected_filings": ["UCC-2024-NY-0012847", "UCC-2024-NY-0015921"]
    }
    result5 = score_task_completion(response5, expected5)
    print(f"\nTest 5 — All expected found + extra:")
    print(f"  Score: {result5['score']} (expected: 1.0)")
    print(f"  Extra: {result5['extra']}")
    assert result5["score"] == 1.0, f"Expected 1.0, got {result5['score']}"

    print("\n" + "=" * 50)
    print("All self-tests passed!")

File: labs/M18-evaluation-testing/solution/task_scorer.js
import { fileURLToPath } from "url";
/**
 * M18 — Task Completion Scorer (Node.js Solution)
 * Scores whether the agent found the correct UCC filings.
 * Uses partial credit: finding 3 of 5 expected filings = 0.6.
 */

/**
 * Extract UCC filing numbers from response text.
 * Pattern: UCC-YYYY-XX-NNNNNNN
 */
function extractFilingNumbers(text) {
  const pattern = /UCC-\d{4}-[A-Z]{2}-\d{7}/g;
  const matches = text.match(pattern) || [];
  return [...new Set(matches)]; // deduplicate
}

/**
 * Score whether the agent found the correct filings.
 *
 * @param {string} response - The agent's text response
 * @param {object} expected - Object with expected_filings array
 * @returns {{ score: number, found: string[], missed: string[], extra: string[], details: string }}
 */
function scoreTaskCompletion(response, expected) {
  const expectedFilings = new Set(expected.expected_filings || []);
  const foundFilings = new Set(extractFilingNumbers(response));

  // Edge case: no filings expected
  if (expectedFilings.size === 0) {
    if (foundFilings.size === 0) {
      return {
        score: 1.0,
        found: [],
        missed: [],
        extra: [],
        details: "Correctly returned no filings (none expected).",
      };
    } else {
      return {
        score: 0.0,
        found: [],
        missed: [],
        extra: [...foundFilings].sort(),
        details: `Expected no filings but found ${foundFilings.size}.`,
      };
    }
  }

  // Calculate matches
  const correctlyFound = [...expectedFilings].filter((f) => foundFilings.has(f));
  const missed = [...expectedFilings].filter((f) => !foundFilings.has(f));
  const extra = [...foundFilings].filter((f) => !expectedFilings.has(f));

  const score = correctlyFound.length / expectedFilings.size;

  const detailParts = [];
  if (correctlyFound.length > 0) {
    detailParts.push(
      `Found ${correctlyFound.length}/${expectedFilings.size} expected filings.`
    );
  }
  if (missed.length > 0) {
    detailParts.push(`Missed: ${missed.sort().join(", ")}.`);
  }
  if (extra.length > 0) {
    detailParts.push(`Extra (not expected): ${extra.sort().join(", ")}.`);
  }
  if (score === 1.0) {
    detailParts.push("Perfect match!");
  }

  return {
    score,
    found: correctlyFound.sort(),
    missed: missed.sort(),
    extra: extra.sort(),
    details: detailParts.join(" "),
  };
}

// ---------------------------------------------------------------------------
// Self-test
// ---------------------------------------------------------------------------
// ESM has no require.main; compare the resolved entry path instead.
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  console.log("Task Completion Scorer — Self-Test");
  console.log("=".repeat(50));

  // Test 1: Perfect match
  const result1 = scoreTaskCompletion(
    "Found 2 filings:\n- UCC-2024-NY-0012847 (Active)\n- UCC-2024-NY-0015921 (Active)",
    { expected_filings: ["UCC-2024-NY-0012847", "UCC-2024-NY-0015921"] }
  );
  console.log(`\nTest 1 — Perfect match: score=${result1.score} (expected: 1.0)`);
  console.assert(result1.score === 1.0, `Expected 1.0, got ${result1.score}`);

  // Test 2: Partial match
  const result2 = scoreTaskCompletion(
    "I found filing UCC-2024-NY-0012847 for Acme Corporation.",
    { expected_filings: ["UCC-2024-NY-0012847", "UCC-2024-NY-0015921"] }
  );
  console.log(`\nTest 2 — Partial match: score=${result2.score} (expected: 0.5)`);
  console.assert(result2.score === 0.5, `Expected 0.5, got ${result2.score}`);

  // Test 3: No match
  const result3 = scoreTaskCompletion(
    "I could not find any filings for that entity.",
    { expected_filings: ["UCC-2024-NY-0012847"] }
  );
  console.log(`\nTest 3 — No match: score=${result3.score} (expected: 0.0)`);
  console.assert(result3.score === 0.0, `Expected 0.0, got ${result3.score}`);

  // Test 4: Empty expected
  const result4 = scoreTaskCompletion("No filings were found for XYZ Corp.", {
    expected_filings: [],
  });
  console.log(`\nTest 4 — Empty expected: score=${result4.score} (expected: 1.0)`);
  console.assert(result4.score === 1.0, `Expected 1.0, got ${result4.score}`);

  console.log("\n" + "=".repeat(50));
  console.log("All self-tests passed!");
}

export { extractFilingNumbers, scoreTaskCompletion };
